presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -85,4 +89,36 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                 node,
                 Assignments.identity(restrictedOutputs)));
     }
+
+    public static List<Set<Symbol>> requiredJoinInputs(JoinNode node, Set<Symbol> requiredJoinOutputs)
+    {
+        Set<Symbol> globallyUsable = requiredJoinOutputs;
+        if (node.getFilter().isPresent()) {
+            globallyUsable = ImmutableSet.<Symbol>builder()
+                    .addAll(requiredJoinOutputs)
+                    .addAll(DependencyExtractor.extractUnique(node.getFilter().get()))
```

Comment:
GroupReference is a PlanNode, not a Node -- so its filter can't be one, as I understand it.